GTK: content scale should never be negative

### Discussed in https://github.com/ghostty-org/ghostty/discussions/5906

<div type='discussions-op-text'>

<sup>Originally posted by **ChausseBenjamin** February 20, 2025</sup>
When building the ghostty binary from source on a gentoo system. I am greeted with a blank screen.
Here are logs printed to stdout when launched from another terminal:
```
info: ghostty version=1.1.3-main+38908e01
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_CA.utf8
info(gtk): GTK version build=4.16.12 runtime=4.16.12
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2
info: reading configuration file path=/home/master/.config/ghostty/config
info(config): default shell source=env value=/bin/zsh
warning(gtk): setting GDK_DEBUG=gl-no-fractional
warning(gtk): setting GDK_DISABLE=gles-api,color-mgmt,vulkan
warning(gtk): setting GSK_RENDERER=opengl
Unrecognized value "color-mgmt". Try GDK_DISABLE=help
info(grid): loaded OpenGL 4.6
info(font_shared_grid_set): font regular: VulfMono-Regular
info(font_shared_grid_set): font bold: VulfMono-Bold
info(font_shared_grid_set): font italic: VulfMono-Italic
info(font_shared_grid_set): font bold_italic: VulfMono-BoldItalic
info(io_exec): found Ghostty resources dir: /home/master/.local/src/ghostty/zig-out/share/ghostty
info(io_exec): shell integration automatically injected shell=termio.shell_integration.Shell.zsh
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(surface): WARNING: very small terminal grid detected with padding set. Is your padding reasonable?
info(io_exec): started subcommand path=/bin/sh pid=24880
info(io_exec): subcommand cgroup=-
warning(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop
info(grid): reallocating GPU buffer old=0 new=8
```

Note that this occurs even on a blank config (no fonts or background). 
Also, the visuals work fine when compiling with the following command (albeit the app is slow then):
```
zig build -Dapp-runtime=glfw run
```

</div>

## Further Discussion

I don't know what causes this, but clearly in the discussion comments someone is getting a negative xscale/yscale from GTK. I can't think of any reason this should ever be allowed and it's causing problems later since we convert these to unsigned integers at some point after doing some math.

I'm not sure what the right behavior is but if a scale is negative, we should probably assume it's invalid and just revert back to `1`?

Interestingly, I can reproduce this on one of my Gentoo systems (similar log when starting ghostty as posted here), but not on another one. I ensured that the way ghoststy is built is the same (-X -adwaita in my case) but the problem persists.  The other 2 differences are: a) different hardware (both Intel but different generations); b) I haven't rebooted the system where it works.
Maybe this helps, ghostty produces this in Sway's log:
```
00:28:47.912 [ERROR] [wlr] [backend/drm/atomic.c:79] connector HDMI-A-1: Atomic commit failed: Device or resource busy
00:28:47.912 [ERROR] [sway/desktop/output.c:317] Page-flip failed on output HDMI-A-1
```
(obviously, I can't see this in the system where it does work)

--- a/src/apprt/gtk/Surface.zig
+++ b/src/apprt/gtk/Surface.zig
@@ -852,8 +852,8 @@ pub fn getContentScale(self: *const Surface) !apprt.ContentScale {
         // Future: detect GTK version 4.12+ and use gdk_surface_get_scale so we
         // can support fractional scaling.
         const scale = widget.getScaleFactor();
-        if (scale < 0) {
-            log.warn("gtk_widget_get_scale_factor returned a negative number: {d:.3}", .{scale});
+        if (scale <= 0) {
+            log.warn("gtk_widget_get_scale_factor returned a non-positive number: {}", .{scale});
             break :scale 1.0;
         }
         break :scale @floatFromInt(scale);

--- a/src/apprt/gtk/Surface.zig
+++ b/src/apprt/gtk/Surface.zig
@@ -847,28 +847,41 @@ pub fn shouldClose(self: *const Surface) bool {
 }
 
 pub fn getContentScale(self: *const Surface) !apprt.ContentScale {
-    // Future: detect GTK version 4.12+ and use gdk_surface_get_scale so we
-    // can support fractional scaling.
-    const gtk_scale: f32 = @floatFromInt(c.gtk_widget_get_scale_factor(@ptrCast(self.gl_area)));
+    const gtk_scale: f32 = scale: {
+        const widget: *gtk.Widget = @ptrCast(@alignCast(self.gl_area));
+        // Future: detect GTK version 4.12+ and use gdk_surface_get_scale so we
+        // can support fractional scaling.
+        const scale = widget.getScaleFactor();
+        if (scale < 0) {
+            log.warn("gtk_widget_get_scale_factor returned a negative number: {d:.3}", .{scale});
+            break :scale 1.0;
+        }
+        break :scale @floatFromInt(scale);
+    };
 
     // Also scale using font-specific DPI, which is often exposed to the user
     // via DE accessibility settings (see https://docs.gtk.org/gtk4/class.Settings.html).
     const xft_dpi_scale = xft_scale: {
         // gtk-xft-dpi is font DPI multiplied by 1024. See
         // https://docs.gtk.org/gtk4/property.Settings.gtk-xft-dpi.html
-        const settings = c.gtk_settings_get_default();
+        const settings = gtk.Settings.getDefault() orelse break :xft_scale 1.0;
+        var value = std.mem.zeroes(gobject.Value);
+        defer value.unset();
+        _ = value.init(gobject.ext.typeFor(c_int));
+        settings.as(gobject.Object).getProperty("gtk-xft-dpi", &value);
+        const gtk_xft_dpi = value.getInt();
 
-        var value: c.GValue = std.mem.zeroes(c.GValue);
-        defer c.g_value_unset(&value);
-        _ = c.g_value_init(&value, c.G_TYPE_INT);
-        c.g_object_get_property(@ptrCast(@alignCast(settings)), "gtk-xft-dpi", &value);
-        const gtk_xft_dpi = c.g_value_get_int(&value);
+        // ensure that the scale is never negative
+        if (gtk_xft_dpi < 0) {
+            log.warn("gtk-xft-dpi setting was negative: {d:.3}", .{gtk_xft_dpi});
+            break :xft_scale 1.0;
+        }
 
         // As noted above gtk-xft-dpi is multiplied by 1024, so we divide by
         // 1024, then divide by the default value (96) to derive a scale. Note
         // gtk-xft-dpi can be fractional, so we use floating point math here.
-        const xft_dpi: f32 = @as(f32, @floatFromInt(gtk_xft_dpi)) / 1024;
-        break :xft_scale xft_dpi / 96;
+        const xft_dpi: f32 = @as(f32, @floatFromInt(gtk_xft_dpi)) / 1024.0;
+        break :xft_scale xft_dpi / 96.0;
     };
 
     const scale = gtk_scale * xft_dpi_scale;

--- a/src/apprt/gtk/Surface.zig
+++ b/src/apprt/gtk/Surface.zig
@@ -871,9 +871,12 @@ pub fn getContentScale(self: *const Surface) !apprt.ContentScale {
         settings.as(gobject.Object).getProperty("gtk-xft-dpi", &value);
         const gtk_xft_dpi = value.getInt();
 
-        // ensure that the scale is never negative
-        if (gtk_xft_dpi < 0) {
-            log.warn("gtk-xft-dpi setting was negative: {d:.3}", .{gtk_xft_dpi});
+        // Use a value of 1.0 for the XFT DPI scale if the setting is <= 0
+        // See:
+        // https://gitlab.gnome.org/GNOME/libadwaita/-/commit/a7738a4d269bfdf4d8d5429ca73ccdd9b2450421
+        // https://gitlab.gnome.org/GNOME/libadwaita/-/commit/9759d3fd81129608dd78116001928f2aed974ead
+        if (gtk_xft_dpi <= 0) {
+            log.warn("gtk-xft-dpi was not set, using default value", .{});
             break :xft_scale 1.0;
         }
 

